test/challtestsrv/challenge-servers.go

```diff
@@ -0,0 +1,123 @@
+package challtestsrv
+
+import (
+	"fmt"
+	"log"
+	"os"
+	"sync"
+)
+
+// ChallSrv is a multi-purpose challenge server. Each ChallSrv may have one or
+// more ACME challenges it provides servers for.
+type ChallSrv struct {
+	log *log.Logger
+	// Shutdown is a channel used to request the challenge server cleanly shut down
+	shutdown chan bool
+
+	// httpOneAddrs are the HTTP-01 challenge server bind address(es)/port(s). If
```

Comment:
It seems a little awkward to copy the config values straight into the `ChallSrv` object. I know we do that in some places, e.g. for timeouts. In this case, I think it would be a little more idiomatic to go ahead and construct the HTTP and DNS server objects in `ChallSrv.New()`, and store those on the returned ChallSrv. Then when `ChallSrv.Run` is called, it can start up the listener loops on those sub-servers. This also lets us switch up the "return cleanup" pattern that `httpOneServer` and `dnsOneServer` have. Since we'd be storing references to the server objects on ChallSrv, we can simply call their Shutdown methods when `ChallSrv.Shutdown()` is called.

Reply:
I think this design would also potentially allow us to eliminate the `shutdown chan bool`.